include/deal.II/base/vectorization.h

```diff
@@ -868,47 +868,42 @@ class VectorizedArray<double>
+  // vectorized array of half the size with 256 bits
   const unsigned int n_chunks = n_entries / 4;
-  for (unsigned int outer = 0; outer < 8; outer += 4)
+  for (unsigned int i = 0; i < n_chunks; ++i)
     {
-      const double *in0 = in + offsets[0 + outer];
-      const double *in1 = in + offsets[1 + outer];
-      const double *in2 = in + offsets[2 + outer];
-      const double *in3 = in + offsets[3 + outer];
-
-      for (unsigned int i = 0; i < n_chunks; ++i)
-        {
-          __m256d u0 = _mm256_loadu_pd(in0 + 4 * i);
-          __m256d u1 = _mm256_loadu_pd(in1 + 4 * i);
-          __m256d u2 = _mm256_loadu_pd(in2 + 4 * i);
-          __m256d u3 = _mm256_loadu_pd(in3 + 4 * i);
-          __m256d t0 = _mm256_permute2f128_pd(u0, u2, 0x20);
-          __m256d t1 = _mm256_permute2f128_pd(u1, u3, 0x20);
-          __m256d t2 = _mm256_permute2f128_pd(u0, u2, 0x31);
-          __m256d t3 = _mm256_permute2f128_pd(u1, u3, 0x31);
-          *reinterpret_cast<__m256d *>(
-            reinterpret_cast<double *>(&out[4 * i + 0].data) + outer) =
-            _mm256_unpacklo_pd(t0, t1);
-          *reinterpret_cast<__m256d *>(
-            reinterpret_cast<double *>(&out[4 * i + 1].data) + outer) =
-            _mm256_unpackhi_pd(t0, t1);
-          *reinterpret_cast<__m256d *>(
-            reinterpret_cast<double *>(&out[4 * i + 2].data) + outer) =
-            _mm256_unpacklo_pd(t2, t3);
-          *reinterpret_cast<__m256d *>(
-            reinterpret_cast<double *>(&out[4 * i + 3].data) + outer) =
-            _mm256_unpackhi_pd(t2, t3);
-        }
-      for (unsigned int i = 4 * n_chunks; i < n_entries; ++i)
-        for (unsigned int v = 0; v < 4; ++v)
-          out[i][outer + v] = in[offsets[v + outer] + i];
+      __m512d t0, t1, t2, t3 = {};
+
+      t0 = _mm512_insertf64x4(t3, _mm256_loadu_pd(in + offsets[0] + 4 * i), 0);
+      t0 = _mm512_insertf64x4(t0, _mm256_loadu_pd(in + offsets[2] + 4 * i), 1);
+      t1 = _mm512_insertf64x4(t3, _mm256_loadu_pd(in + offsets[1] + 4 * i), 0);
+      t1 = _mm512_insertf64x4(t1, _mm256_loadu_pd(in + offsets[3] + 4 * i), 1);
+      t2 = _mm512_insertf64x4(t3, _mm256_loadu_pd(in + offsets[4] + 4 * i), 0);
+      t2 = _mm512_insertf64x4(t2, _mm256_loadu_pd(in + offsets[6] + 4 * i), 1);
+      t3 = _mm512_insertf64x4(t3, _mm256_loadu_pd(in + offsets[5] + 4 * i), 0);
+      t3 = _mm512_insertf64x4(t3, _mm256_loadu_pd(in + offsets[7] + 4 * i), 1);
+
+      __m512d v0          = _mm512_shuffle_f64x2(t0, t2, 0x88);
+      __m512d v1          = _mm512_shuffle_f64x2(t0, t2, 0xdd);
+      __m512d v2          = _mm512_shuffle_f64x2(t1, t3, 0x88);
+      __m512d v3          = _mm512_shuffle_f64x2(t1, t3, 0xdd);
+      out[4 * i + 0].data = _mm512_unpacklo_pd(v0, v2);
+      out[4 * i + 1].data = _mm512_unpackhi_pd(v0, v2);
+      out[4 * i + 2].data = _mm512_unpacklo_pd(v1, v3);
+      out[4 * i + 3].data = _mm512_unpackhi_pd(v1, v3);
     }
+  for (unsigned int i = 4 * n_chunks; i < n_entries; ++i)
```

Comment:
here (and in other places): can you please leave a comment, that these are the "left-over" pieces done manually?

Reply:
Yes.